apps/src/gamelab/GameLab.js

```diff
@@ -0,0 +1,574 @@
+    // and analyze success/failure based on what was drawn.
+    // Otherwise, abnormal termination is a user error.
+    if (e !== Infinity) {
+      // call window.onerror so that we get new relic collection.  prepend with
+      // UserCode so that it's clear this is in eval'ed code.
+      if (window.onerror) {
+        window.onerror("UserCode:" + e.message, document.URL, 0);
+      }
+      window.alert(e);
+    }
+  }
+};
+
+/**
+ * Set up this.code, this.interpreter, etc. to run code for editCode levels
+ */
+
+/*
+GameLab.prototype.generateTurtleCodeFromJS_ = function () {
+  this.code = dropletUtils.generateCodeAliases(dropletConfig, 'Turtle');
+  this.userCodeStartOffset = this.code.length;
+  this.code += this.studioApp_.editor.getValue();
+  this.userCodeLength = this.code.length - this.userCodeStartOffset;
+
+  var session = this.studioApp_.editor.aceEditor.getSession();
+  this.cumulativeLength = codegen.aceCalculateCumulativeLength(session);
+
+  var initFunc = _.bind(function(interpreter, scope) {
+    codegen.initJSInterpreter(interpreter, null, null, scope, {
+      Turtle: this.api
+    });
+  }, this);
+  this.interpreter = new window.Interpreter(this.code, initFunc);
+};
+*/
+
+/**
+ * Execute the user's code.  Heaven help us...
+ */
+GameLab.prototype.execute = function() {
+  // Reset all state.
+  this.studioApp_.reset();
+
+  if (this.studioApp_.isUsingBlockly() &&
+      (this.studioApp_.hasExtraTopBlocks() ||
+        this.studioApp_.hasDuplicateVariablesInForLoops())) {
+    // immediately check answer, which will fail and report top level blocks
+    this.checkAnswer();
+    return;
+  }
+
+  if (this.level.editCode) {
+    this.JSInterpreter = new JSInterpreter({
+      code: this.studioApp_.getCode(),
+      blocks: dropletConfig.blocks,
+      blockFilter: this.level.executePaletteApisOnly && this.level.codeFunctions,
+      enableEvents: true,
+      studioApp: this.studioApp_,
+      onExecutionError: _.bind(this.handleExecutionError, this),
+    });
+    if (!this.JSInterpreter.initialized()) {
+      return;
+    }
+    /*
+    if (this.checkForEditCodePreExecutionFailure()) {
+      return this.onPuzzleComplete();
+    }
+    */
+  } else {
+    this.code = Blockly.Generator.blockSpaceToCode('JavaScript');
+    this.evalCode(this.code);
+  }
+
+  // api.log now contains a transcript of all the user's actions.
+  this.studioApp_.playAudio('start', {loop : true});
+  // animate the transcript.
+
+  if (this.studioApp_.isUsingBlockly()) {
+    // Disable toolbox while running
+    Blockly.mainBlockSpaceEditor.setEnableToolbox(false);
+  }
+
+  this.tickIntervalId = window.setInterval(_.bind(this.onTick, this), 33);
+};
+
+GameLab.prototype.onTick = function () {
+  this.tickCount++;
+  var i;
+
+  if (this.tickCount === 1) {
+    this.callHandler('whenGameStarts');
+  }
+
+  this.executeQueue('whenGameStarts');
+
+  this.callHandler('repeatForever');
+  this.executeQueue('repeatForever');
```

Comment:
Regarding the first issue, yes, the `repeatForever` code does start running concurrently with the `whenGameStarts` code for blockly levels if you place a long-running block under the "when run" block (such as "wait", "showTitleScreen", "saySprite", "throwProjectile", or "moveDistance").

For Play Lab, this could either good or bad. For new apps, I'm not sure that we want to introduce these command queues.